pkg/apps/strategy/support/lifecycle_hooks.go

```diff
@@ -0,0 +1,285 @@
+	appsutil "github.com/openshift/origin/pkg/apps/util"
+)
+
+type HookExecutor interface {
+	RunHook(hook *appsapi.LifecycleHook, controller *kapi.ReplicationController, stage string, timeout time.Duration) error
+}
+
+// LifecycleHookController operates the execution and lifecycle of deployment lifecycle hooks.
+type LifecycleHookController struct {
+	podClient  coreclient.PodsGetter
+	kubeClient kubernetes.Interface
+
+	recorder record.EventRecorder
+	out      io.Writer
+}
+
+var (
+	HookPodTimeoutError     error
+	DeployerPodTimeoutError error
+)
+
+func NewLifecycleHookController(kubeclientset kubernetes.Interface, out io.Writer) *LifecycleHookController {
+	c := &LifecycleHookController{
+		kubeClient: kubeclientset,
+		out:        out,
+	}
+
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(glog.Infof)
+	eventBroadcaster.StartRecordingToSink(&typedcorev1.EventSinkImpl{Interface: kubeclientset.CoreV1().Events("")})
+
+	c.recorder = eventBroadcaster.NewRecorder(scheme.Scheme, corev1.EventSource{Component: "deployer"})
+
+	return c
+}
+
+type lifecycleHook struct {
+	podLister corelisters.PodLister
+	podClient typedcorev1.PodInterface
+
+	rcName      string
+	rcNamespace string
+
+	deployerStartTime     *time.Time
+	deployerStartTimeChan chan time.Time
+
+	hookStartTime     *time.Time
+	hookStartTimeChan chan time.Time
+
+	hookPodName     string
+	hookPodRestarts int32
+
+	hookCompleteChan chan struct{}
+
+	logStreamStarted      bool
+	logStreamCompleteChan chan struct{}
+
+	stage string
+	out   io.Writer
+}
+
+func (c *LifecycleHookController) RunHook(hook *appsapi.LifecycleHook, rc *kapi.ReplicationController, stage string, timeout time.Duration) error {
+	// Start new filtered informer for all deployer managed pods
+	informer := coreinformers.NewFilteredPodInformer(c.kubeClient, rc.Namespace, 60*time.Second, cache.Indexers{}, func(options *metav1.ListOptions) {
+		options.LabelSelector = labels.SelectorFromSet(map[string]string{appsapi.DeployerPodForDeploymentLabel: rc.Name}).String()
+	})
+
+	o := &lifecycleHook{
+		rcName:      rc.Name,
+		rcNamespace: rc.Namespace,
+		podLister:   corelisters.NewPodLister(informer.GetIndexer()),
+		podClient:   c.kubeClient.CoreV1().Pods(rc.Namespace),
+		out:         c.out,
+		stage:       stage,
+		deployerStartTimeChan: make(chan time.Time),
+		hookStartTimeChan:     make(chan time.Time),
+		hookCompleteChan:      make(chan struct{}),
+		logStreamCompleteChan: make(chan struct{}),
+	}
+
+	informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    o.addPod,
+		UpdateFunc: o.updatePod,
+		DeleteFunc: o.deletePod,
+	})
+
+	stopChan := make(chan struct{})
+
+	go informer.Run(stopChan)
+	if !cache.WaitForCacheSync(stopChan, informer.HasSynced) {
+		return o.error("wait for cache sync failed")
+	}
+
+	defer close(stopChan)
+
+	// Block until we know the the deployer pod started
+	select {
+	case t := <-o.deployerStartTimeChan:
```

Comment:
@tnozicka block this kubelet make the update, if kubelet is not making the update, lets fix kubelet instead of workarounding this with time.Now()